proposals/NNNN-hlsl-triangle-object-positions.md

```diff
@@ -0,0 +1,299 @@
+<!-- {% raw %} -->
+
+# HLSL TriangleObjectPositions
+
+* Proposal: [NNNN](NNNN-triangle-object-positions.md)
+* Author(s): [Tex Riddell](https://github.com/tex3d)
+* Sponsor: [Tex Riddell](https://github.com/tex3d)
+* Status: **Under Consideration**
+
+---
+
+## Introduction
+
+This proposal adds intrinsics that can be called from an Any hit or Closest
+hit shader to obtain the positions of the vertices for the triangle that has
+been hit.
+
+---
+
+## Motivation
+
+Developers often need to know the positions of the vertices for the triangle
+that was hit in an Any/Closest hit shader. Currently, the only way to do this is
+for the developer to provide separate buffers containing this data. This
+information is also available in the acceleration structure, but there is
+currently no way to access it.
+
+If developers can access this data from their shader code then this will remove
+the need have a duplicate copy of it. In addition, this will provide drivers
+with opportunities to optimize the data layout for their implementations.
+
+---
+
+## Proposed solution
+
+Add intrinsics to look up the object-space vertex positions of the triangle for
+the current hit, or RayQuery candidate/committed hits.
+
+For context, see related sections in DirectX Raytracing Specification:
+[TriangleObjectPositions][dxr-tri-obj-pos],
+[RayQuery::CandidateTriangleObjectPositions][dxr-rq-can-tri-obj-pos],
+[RayQuery::CommittedTriangleObjectPositions][dxr-rq-com-tri-obj-pos].
+
+---
+
+## Detailed design
+
+---
+
+### HLSL Additions
+
+A new built-in structure for returning all three object-space triangle
+positions is added: `BuiltInTrianglePositions`.
+
+> See Issue [Return Type](#return-type)
+
+```cpp
+/// \brief Stores the positions for all vertices of a triangle
+struct BuiltInTrianglePositions {
+  float3 p0, p1, p2;
+};
+```
+
+A new intrinsic function to retrieve triangle object-space positions for the
+currently hit triangle: `TriangleObjectPositions`.
+
+> See Issue [Intrinsic Naming](#intrinsic-naming)
+
+```cpp
+/// \brief Retrieve current hit triangle object-space vertex positions
+/// \returns position of vertex in object-space
+///
+/// Minimum shader model: (Version TBD)
+/// Allowed shader types: anyhit, closesthit.
+///
+/// Hit kind must be a triangle, where HitKind() returns either
+/// HIT_KIND_TRIANGLE_FRONT_FACE (254) or HIT_KIND_TRIANGLE_BACK_FACE (255),
+/// otherwise undefined behavior results.
+///
+/// Ordering of positions returned is based on the order used to build the
+/// acceleration structure, and must match barycentric ordering.
+BuiltInTrianglePositions TriangleObjectPositions();
+```
+
+Two new intrinsic methods on `RayQuery` to retrieve triangle object-space
+positions for the candidate or committed triangle hits:
+`CandidateTriangleObjectPositions`, and `CommittedTriangleObjectPositions`.
+
+```cpp
+template<uint ConstRayFlags>
+class RayQuery {
```

Comment:
It will be simpler for us to comply with our namespacing policy (https://github.com/microsoft/hlsl-specs/blob/main/docs/Process.md#conforming-extensions) if we also provide this implementation for SPIRV.

I believe this is fully compatible with [SPV_KHR_ray_tracing_position_fetch](https://github.khronos.org/SPIRV-Registry/extensions/KHR/SPV_KHR_ray_tracing_position_fetch.html), but it would be good to get an opinion from @s-perron and/or @Keenuts.